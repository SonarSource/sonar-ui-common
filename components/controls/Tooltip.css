/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
.tooltip {
  position: absolute;
  z-index: var(--tooltipZIndex);
  display: block;
  height: auto;
  box-sizing: border-box;
  font-size: var(--smallFontSize);
  font-weight: 300;
  line-height: 1.5;
  animation: fadeIn 0.3s forwards;
}

.tooltip.top {
  padding: 5px 0;
  margin-top: -3px;
}

.tooltip.right {
  padding: 0 5px;
  margin-left: 3px;
}

.tooltip.bottom {
  padding: 5px 0;
  margin-top: 3px;
}

.tooltip.left {
  padding: 0 5px;
  margin-left: -3px;
}

.tooltip-inner {
  max-width: 300px;
  text-align: left;
  text-decoration: none;
  border-radius: 4px;
  overflow: hidden;
  word-break: break-word;
}

.tooltip-inner {
  padding: 12px 17px;
  color: #fff;
  background-color: #475760;
  letter-spacing: 0.04em;
}

.tooltip-inner .alert {
  margin-bottom: 5px;
  border-radius: 4px;
}

.tooltip-inner a {
  border-bottom-color: #8da6b3;
  color: #a5d0ea;
}

.tooltip-inner hr {
  background-color: #5d6d75;
}

.tooltip-arrow {
  position: absolute;
  width: 0;
  height: 0;
  border: solid transparent;
}

.tooltip.top .tooltip-arrow {
  bottom: 0;
  left: 50%;
  border-width: 5px 5px 0;
  transform: translateX(-5px);
  border-top-color: #475760;
}

.tooltip.right .tooltip-arrow {
  top: 50%;
  left: 0;
  transform: translateY(-5px);
  border-width: 5px 5px 5px 0;
  border-right-color: #475760;
}

.tooltip.left .tooltip-arrow {
  top: 50%;
  right: 0;
  transform: translateY(-5px);
  border-width: 5px 0 5px 5px;
  border-left-color: #475760;
}

.tooltip.bottom .tooltip-arrow {
  top: 0;
  left: 50%;
  transform: translateX(-5px);
  border-width: 0 5px 5px;
  border-bottom-color: #475760;
}

/* Workaround for react issue with onMouseLeave in disabled buttons: https://github.com/facebook/react/issues/4251 */
.tooltip button[disabled] {
  pointer-events: none;
}

@keyframes fadeIn {
  from {
    opacity: 0;
  }

  to {
    opacity: 1;
  }
}
