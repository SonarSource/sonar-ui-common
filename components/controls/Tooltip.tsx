/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { createPortal, findDOMNode } from 'react-dom';
import { throttle } from 'lodash';
import ScreePositionFixer from './ScreenPositionFixer';
import './Tooltip.css';

export type Placement = 'bottom' | 'right' | 'left' | 'top';

interface Props {
  classNameSpace?: string;
  children: React.ReactElement<{}>;
  mouseEnterDelay?: number;
  mouseLeaveDelay?: number;
  onShow?: () => void;
  onHide?: () => void;
  overlay: React.ReactNode;
  placement?: Placement;
  visible?: boolean;
}

interface Measurements {
  height: number;
  left: number;
  top: number;
  width: number;
}

interface OwnState {
  visible: boolean;
}

type State = OwnState & Partial<Measurements>;

function isMeasured(state: State): state is OwnState & Measurements {
  return state.height !== undefined;
}

export default function Tooltip(props: Props) {
  // overlay is a ReactNode, so it can be `undefined` or `null`
  // this allows to easily render a tooltip conditionally
  // more generaly we avoid rendering empty tooltips
  return props.overlay != null && props.overlay !== '' ? (
    <TooltipInner {...props} />
  ) : (
    props.children
  );
}

export class TooltipInner extends React.Component<Props, State> {
  throttledPositionTooltip: () => void;
  mouseEnterTimeout?: number;
  mouseLeaveTimeout?: number;
  tooltipNode?: HTMLElement | null;
  mounted = false;
  mouseIn = false;

  static defaultProps = {
    mouseEnterDelay: 0.1
  };

  constructor(props: Props) {
    super(props);
    this.state = {
      visible: props.visible !== undefined ? props.visible : false
    };
    this.throttledPositionTooltip = throttle(this.positionTooltip, 10);
  }

  componentDidMount() {
    this.mounted = true;
    if (this.props.visible === true) {
      this.positionTooltip();
      this.addEventListeners();
    }
  }

  componentDidUpdate(prevProps: Props, prevState: State) {
    if (
      // opens
      (this.props.visible === true && prevProps.visible === false) ||
      (this.props.visible === undefined &&
        this.state.visible === true &&
        prevState.visible === false)
    ) {
      this.positionTooltip();
      this.addEventListeners();
    } else if (
      // closes
      (this.props.visible === false && prevProps.visible === true) ||
      (this.props.visible === undefined &&
        this.state.visible === false &&
        prevState.visible === true)
    ) {
      this.clearPosition();
      this.removeEventListeners();
    }
  }

  componentWillUnmount() {
    this.mounted = false;
    this.removeEventListeners();
    this.clearTimeouts();
  }

  addEventListeners = () => {
    window.addEventListener('resize', this.throttledPositionTooltip);
    window.addEventListener('scroll', this.throttledPositionTooltip);
  };

  removeEventListeners = () => {
    window.removeEventListener('resize', this.throttledPositionTooltip);
    window.removeEventListener('scroll', this.throttledPositionTooltip);
  };

  clearTimeouts = () => {
    window.clearTimeout(this.mouseEnterTimeout);
    window.clearTimeout(this.mouseLeaveTimeout);
  };

  isVisible = () => {
    return this.props.visible !== undefined ? this.props.visible : this.state.visible;
  };

  getPlacement = (): Placement => {
    return this.props.placement || 'bottom';
  };

  tooltipNodeRef = (node: HTMLElement | null) => {
    this.tooltipNode = node;
  };

  positionTooltip = () => {
    // `findDOMNode(this)` will search for the DOM node for the current component
    // first it will find a React.Fragment (see `render`),
    // so it will get the DOM node of the first child, i.e. DOM node of `this.props.children`
    // docs: https://reactjs.org/docs/refs-and-the-dom.html#exposing-dom-refs-to-parent-components

    // eslint-disable-next-line react/no-find-dom-node
    const toggleNode = findDOMNode(this);

    if (toggleNode && toggleNode instanceof Element && this.tooltipNode) {
      const toggleRect = toggleNode.getBoundingClientRect();
      const tooltipRect = this.tooltipNode.getBoundingClientRect();
      const { width, height } = tooltipRect;

      let left = 0;
      let top = 0;

      switch (this.getPlacement()) {
        case 'bottom':
          left = toggleRect.left + toggleRect.width / 2 - width / 2;
          top = toggleRect.top + toggleRect.height;
          break;
        case 'top':
          left = toggleRect.left + toggleRect.width / 2 - width / 2;
          top = toggleRect.top - height;
          break;
        case 'right':
          left = toggleRect.left + toggleRect.width;
          top = toggleRect.top + toggleRect.height / 2 - height / 2;
          break;
        case 'left':
          left = toggleRect.left - width;
          top = toggleRect.top + toggleRect.height / 2 - height / 2;
          break;
      }

      // save width and height (and later set in `render`) to avoid resizing the tooltip element,
      // when it's placed close to the window edge
      const measurements: Measurements = {
        left: window.pageXOffset + left,
        top: window.pageYOffset + top,
        width,
        height
      };
      this.setState(measurements);
    }
  };

  clearPosition = () => {
    this.setState({
      left: undefined,
      top: undefined,
      width: undefined,
      height: undefined
    });
  };

  handleMouseEnter = () => {
    this.mouseEnterTimeout = window.setTimeout(() => {
      // for some reason even after the `this.mouseEnterTimeout` is cleared, it still triggers
      // to workaround this issue, check that its value is not `undefined`
      // (if it's `undefined`, it means the timer has been reset)
      if (
        this.mounted &&
        this.props.visible === undefined &&
        this.mouseEnterTimeout !== undefined
      ) {
        this.setState({ visible: true });
      }
    }, (this.props.mouseEnterDelay || 0) * 1000);

    if (this.props.onShow) {
      this.props.onShow();
    }
  };

  handleMouseLeave = () => {
    if (this.mouseEnterTimeout !== undefined) {
      window.clearTimeout(this.mouseEnterTimeout);
      this.mouseEnterTimeout = undefined;
    }

    if (!this.mouseIn) {
      this.mouseLeaveTimeout = window.setTimeout(() => {
        if (this.mounted && this.props.visible === undefined && !this.mouseIn) {
          this.setState({ visible: false });
        }
      }, (this.props.mouseLeaveDelay || 0) * 1000);

      if (this.props.onHide) {
        this.props.onHide();
      }
    }
  };

  handleOverlayMouseEnter = () => {
    this.mouseIn = true;
  };

  handleOverlayMouseLeave = () => {
    this.mouseIn = false;
    this.handleMouseLeave();
  };

  render() {
    const { classNameSpace = 'tooltip' } = this.props;
    return (
      <>
        {React.cloneElement(this.props.children, {
          onMouseEnter: this.handleMouseEnter,
          onMouseLeave: this.handleMouseLeave
        })}
        {this.isVisible() && (
          <TooltipPortal>
            <ScreePositionFixer ready={isMeasured(this.state)}>
              {({ leftFix = 0, topFix = 0 }) => (
                <div
                  className={`${classNameSpace} ${this.getPlacement()}`}
                  onMouseEnter={this.handleOverlayMouseEnter}
                  onMouseLeave={this.handleOverlayMouseLeave}
                  ref={this.tooltipNodeRef}
                  style={
                    isMeasured(this.state)
                      ? {
                          left: this.state.left + leftFix,
                          top: this.state.top + topFix,
                          width: this.state.width,
                          height: this.state.height
                        }
                      : undefined
                  }>
                  <div className={`${classNameSpace}-inner`}>{this.props.overlay}</div>
                  <div
                    className={`${classNameSpace}-arrow`}
                    style={
                      isMeasured(this.state)
                        ? { marginLeft: -leftFix, marginTop: -topFix }
                        : undefined
                    }
                  />
                </div>
              )}
            </ScreePositionFixer>
          </TooltipPortal>
        )}
      </>
    );
  }
}

class TooltipPortal extends React.Component {
  el: HTMLElement;

  constructor(props: {}) {
    super(props);
    this.el = document.createElement('div');
  }

  componentDidMount() {
    document.body.appendChild(this.el);
  }

  componentWillUnmount() {
    document.body.removeChild(this.el);
  }

  render() {
    return createPortal(this.props.children, this.el);
  }
}
