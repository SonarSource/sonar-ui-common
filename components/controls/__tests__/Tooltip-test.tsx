/*
 * Sonar UI Common
 * Copyright (C) 2019-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import Tooltip, { TooltipInner } from '../Tooltip';

jest.useFakeTimers();
jest.mock('react-dom', () => {
  const actual = require.requireActual('react-dom');
  return Object.assign({}, actual, {
    findDOMNode: () => undefined
  });
});

it('should render', () => {
  expect(
    shallow(
      <TooltipInner overlay={<span id="overlay" />} visible={false}>
        <div id="tooltip" />
      </TooltipInner>
    )
  ).toMatchSnapshot();
  expect(
    shallow(
      <TooltipInner overlay={<span id="overlay" />} visible={true}>
        <div id="tooltip" />
      </TooltipInner>,
      { disableLifecycleMethods: true }
    )
  ).toMatchSnapshot();
});

it('should open & close', () => {
  const onShow = jest.fn();
  const onHide = jest.fn();
  const wrapper = shallow(
    <TooltipInner onHide={onHide} onShow={onShow} overlay={<span id="overlay" />}>
      <div id="tooltip" />
    </TooltipInner>
  );
  wrapper.find('#tooltip').simulate('mouseenter');
  jest.runOnlyPendingTimers();
  wrapper.update();
  expect(wrapper.find('TooltipPortal').exists()).toBe(true);
  expect(onShow).toBeCalled();

  wrapper.find('#tooltip').simulate('mouseleave');
  jest.runOnlyPendingTimers();
  wrapper.update();
  expect(wrapper.find('TooltipPortal').exists()).toBe(false);
  expect(onHide).toBeCalled();
});

it('should not open when mouse goes away quickly', () => {
  const onShow = jest.fn();
  const onHide = jest.fn();
  const wrapper = shallow(
    <TooltipInner onHide={onHide} onShow={onShow} overlay={<span id="overlay" />}>
      <div id="tooltip" />
    </TooltipInner>
  );

  wrapper.find('#tooltip').simulate('mouseenter');
  wrapper.find('#tooltip').simulate('mouseleave');
  jest.runOnlyPendingTimers();
  wrapper.update();

  expect(wrapper.find('TooltipPortal').exists()).toBe(false);
});

it('should not render tooltip without overlay', () => {
  const wrapper = shallow(
    <Tooltip overlay={undefined}>
      <div id="tooltip" />
    </Tooltip>
  );
  expect(wrapper.type()).toBe('div');
});

it('should not render empty tooltips', () => {
  expect(
    shallow(
      <Tooltip overlay={undefined} visible={true}>
        <div id="tooltip" />
      </Tooltip>
    )
  ).toMatchSnapshot();
  expect(
    shallow(
      <Tooltip overlay="" visible={true}>
        <div id="tooltip" />
      </Tooltip>
    )
  ).toMatchSnapshot();
});
