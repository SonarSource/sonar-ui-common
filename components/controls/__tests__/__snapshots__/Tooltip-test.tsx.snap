// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should not render empty tooltips 1`] = `
<div
  id="tooltip"
/>
`;

exports[`should not render empty tooltips 2`] = `
<div
  id="tooltip"
/>
`;

exports[`should render 1`] = `
<Fragment>
  <div
    id="tooltip"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  />
</Fragment>
`;

exports[`should render 2`] = `
<Fragment>
  <div
    id="tooltip"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  />
  <TooltipPortal>
    <ScreenPositionFixer
      ready={false}
    >
      <Component />
    </ScreenPositionFixer>
  </TooltipPortal>
</Fragment>
`;
